cpp/src/Slice/Grammar.y

```diff
@@ -584,94 +605,14 @@ type_id
 // ----------------------------------------------------------------------
 optional
 // ----------------------------------------------------------------------
-: ICE_OPTIONAL_OPEN ICE_INTEGER_LITERAL ')'
+: ICE_OPTIONAL_OPEN integer_constant ')'
 {
-    auto i = dynamic_pointer_cast<IntegerTok>($2);
-
-    int tag;
-    if (i->v < 0 || i->v > std::numeric_limits<int32_t>::max())
-    {
-        currentUnit->error("tag is out of range");
-        tag = -1;
-    }
-    else
-    {
-        tag = static_cast<int>(i->v);
-    }
-
-    auto m = make_shared<OptionalDefTok>(tag);
-    $$ = m;
-}
-| ICE_OPTIONAL_OPEN scoped_name ')'
-{
-    auto scoped = dynamic_pointer_cast<StringTok>($2);
-    ContainerPtr cont = currentUnit->currentContainer();
-    ContainedList cl = cont->lookupContained(scoped->v, false);
-    if (cl.empty())
-    {
-        EnumeratorList enumerators = cont->enumerators(scoped->v);
-        if (enumerators.size() == 1)
-        {
-            // Found
-            cl.push_back(enumerators.front());
-            scoped->v = enumerators.front()->scoped();
-        }
-        else if (enumerators.size() > 1)
-        {
-            ostringstream os;
-            os << "enumerator '" << scoped->v << "' could designate";
-            bool first = true;
-            for (const auto& p : enumerators)
-            {
-                if (first)
-                {
-                    first = false;
-                }
-                else
-                {
-                    os << " or";
-                }
-
-                os << " '" << p->scoped() << "'";
-            }
-            currentUnit->error(os.str());
-        }
-        else
-        {
-            currentUnit->error(string("'") + scoped->v + "' is not defined");
-        }
-    }
-
-    if (cl.empty())
-    {
-        YYERROR; // Can't continue, jump to next yyerrok
-    }
-    cont->checkIntroduced(scoped->v);
-
+    auto integer = dynamic_pointer_cast<IntegerTok>($2);
     int tag = -1;
-    auto enumerator = dynamic_pointer_cast<Enumerator>(cl.front());
-    auto constant = dynamic_pointer_cast<Const>(cl.front());
-    if (constant)
-    {
-        auto b = dynamic_pointer_cast<Builtin>(constant->type());
-        if (b && b->isIntegralType())
-        {
-            int64_t l = std::stoll(constant->value(), nullptr, 0);
-            if (l < 0 || l > std::numeric_limits<int32_t>::max())
-            {
-                currentUnit->error("tag is out of range");
-            }
-            tag = static_cast<int>(l);
-        }
-    }
-    else if (enumerator)
-    {
-        tag = enumerator->value();
-    }
 
-    if (tag < 0)
```

Comment:
We also had duplicate code checking the bounds of integers, that I centralized in a helper function `checkIntegerBounds`.